File: AdaBoost/myBEMBoost.R
BEMBoost<-function( x, y, x.test=NULL, method='mlp', maxIter=100, BEM=0.5,... ){
	# Feely 2000
	# Big Error Margin Boosting Algorithm
	# Given initialTrainingSet, initialDistribution and BEM
	# D0[i] = initialDistribution[i] for all i
	# t = 0; errCount = 0 
	# 1. Populate currentTrainingSet using Dt
	# 2. Construct network ht and train it using currentTrainingSet.
	# 3. For each member of initialTrainingSet:
	# 	Present that element to the trained network and record the prediction error.
	# If the resulting prediction error is greater then BEM,
	# mark that element as incorrectly predicted.
	# errCount = errCount + 1
	# Otherwise, mark that element as correctly predicted.
	# 4. Calculate the UpFactor and DownFactor as follows:
	# 	errCount
	# initialTrainingSet UpFactor =
	# 	UpFactor
	# DownFactor 1 =
	# 	5. Cycle through each member of the originalTrainingSet and update each element’s
	# distribution in the next training set as follows:
	# 	If the element was incorrectly predicted by the current network:
	# 	distribution [] [] i distribution i UpFactor t+1 = t ×
	# If the element was correctly predicted by the current network:
	# 	distribution [] [] i distribution i DownFactor t+1 = t ×
	# 6. Populate a new training set according to the distribution just calculated.
	# 7. Repeat steps 1-6 until a given maximum number of networks has been created. 
	
	m<-nrow(x)
	n<-ncol(x)
	y<-matrix(y,nrow=m)
	
	T=maxIter # max iter
	#initialize weight distr and error
	epsilon<-beta<-rep(NA,T)
	upfactor<-downfactor<-rep(NA,T)
	fit<-e<-D<-array(NA,dim=c(T,m))
	D[1,]=rep(1/m, m)
	model<-list()
	# iterate
	errorcnt<-rep(0,T)
	
	# iterate
	t=1
	while (t<T){
		# bootstrap
		idx<-sample(m,replace=TRUE,prob=D[t,])
		tth_x=x[idx,]
		tth_y=y[idx,]
		# call weak learner
		model[[t]]=eval(parse(text=method))(tth_x,tth_y,...)
		
		# getback to hypothesis
		fit[t,]<-y_fit<-predict(model[[t]],x)
		AE=abs(y_fit-y) # absolute error
		# calculate error count AE>BEM
		is_AE_gt_BEM=AE>BEM
		errorcnt[t]=sum(is_AE_gt_BEM)
		# check if errorcnt is zero
		if(errorcnt[t]==0)
			break
		# calculate upfactor and downfactor
		upfactor[t]=m/errorcnt[t]
		downfactor[t]=1/upfactor[t]
		
		# update distr
		is_AE_le_BEM=!is_AE_gt_BEM
		
		D[t+1,is_AE_gt_BEM]=D[t,is_AE_gt_BEM]*upfactor[t]
		D[t+1,is_AE_le_BEM]=D[t,is_AE_le_BEM]*downfactor[t]
		# normalize
		D[t+1,]=D[t+1,]/sum(D[t+1,])
		t=t+1
	}
	max_t=t-1
	# final model:
	
	if(max_t>1)
		aggregated.fit<-apply(fit[1:max_t,],2,mean)
	else
		aggregated.fit<-fit[1:max_t,]
	
	# predict
	if(!is.null(x.test)){ # have test dataset
		m.test<-nrow(x.test)
		predict<-array(NA,dim=c(max_t,m.test))
		for (t in 1:max_t){
			predict[t,]<-predict(model[[t]],x.test)
		}
		
		if(max_t>1)
			aggregated.predict<-apply(predict,2,mean)
		else
			aggregated.fit<-predict
		
	}else{
		aggregated.predict=NULL
	}
	return(list(aggregated.fit=aggregated.fit,aggregated.predict=aggregated.predict,
							base.fit=fit,base.predict=predict,max.t=max_t))
}

File: matlab-src/myBEMBoost.m
function [ z_pred ] = myBEMBoost( trn_fun, tst_fun, x, y, z, maxIter, BEM )
% Feely 2000

[m,n]=size(x);
%initialize weight distr and error
D_new=repmat(1/m, m, 1);

T=maxIter; % max iter
% phi=0.5; % error threshold

errorcnt=zeros(1,T);
beta=ones(1,T);
f=cell(1,T);
% iterate
t=1;
while t<=T
    D=D_new;
    % bootstrap with dist D
    [tmp,bootsam]=bootstrp(1, @mean, x, 'Weights',D);
    tth_x=x(bootsam,:);
    tth_y=y(bootsam,:);
    % call weak learner, CART
    f{t}=trn_fun(tth_x,tth_y);
    % getback to hypothesis
    y_fit=tst_fun(f{t},x);
    AE=abs(y_fit-y); % absolute error
    % calculate error count AE>BEM
    is_ae_lt_BEM=AE>BEM;
    errorcnt(t)=sum(is_ae_lt_BEM);
    % calculate upfactor and downfactor
    upfactor(t)=m/errorcnt(t);
    downfactor(t)=1/upfactor(t);
%     
%     if(e(t)==0) % perfect bypass this one
%         continue;
%     end
    
    % update distr
    is_ae_le_BEM=not(is_ae_lt_BEM);
    
    D_new=D.*(is_ae_lt_BEM.*upfactor(t)+is_ae_le_BEM.*downfactor(t));
    % normalize
    D_new=D_new./sum(D_new);
    t=t+1;
end
max_t=t-1;
% final model: cart, beta(t)
% ww=log(1./beta(1:max_t));

% test
pred=zeros(size(z,1), max_t);
for t=1:max_t
    pred(:,t)=tst_fun(f{t},z);
end


z_pred=median(pred,2);
end



